group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,454 @@
+            OptionalLong.empty() : OptionalLong.of(currentTimeMs + request.retentionTimeMs());
+        final OffsetCommitResponseData response = new OffsetCommitResponseData();
+        final List<Record> records = new ArrayList<>();
+
+        request.topics().forEach(topic -> {
+            final OffsetCommitResponseData.OffsetCommitResponseTopic topicResponse =
+                new OffsetCommitResponseData.OffsetCommitResponseTopic()
+                    .setName(topic.name());
+            response.topics().add(topicResponse);
+
+            topic.partitions().forEach(partition -> {
+                if (partition.committedMetadata() != null && partition.committedMetadata().length() > offsetMetadataMaxSize) {
+                    topicResponse.partitions().add(new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.OFFSET_METADATA_TOO_LARGE.code()));
+                } else {
+                    log.debug("[GroupId {}] Committing offsets {} for partition {}-{} from member {} with leader epoch {}.",
+                        request.groupId(), partition.committedOffset(), topic.name(), partition.partitionIndex(),
+                        request.memberId(), partition.committedLeaderEpoch());
+
+                    topicResponse.partitions().add(new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.NONE.code()));
+
+                    final OptionalInt leaderEpoch = partition.committedLeaderEpoch() == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
+                        OptionalInt.empty() : OptionalInt.of(partition.committedLeaderEpoch());
+                    final String metadata = partition.committedMetadata() == null ?
+                        OffsetAndMetadata.NO_METADATA : partition.committedMetadata();
+                    final long commitTimestampMs = partition.commitTimestamp() == OffsetCommitRequest.DEFAULT_TIMESTAMP ?
+                        currentTimeMs : partition.commitTimestamp();
+
+                    final OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(
+                        partition.committedOffset(),
+                        leaderEpoch,
+                        metadata,
+                        commitTimestampMs,
+                        expireTimestampMs
+                    );
+
+                    records.add(RecordHelpers.newOffsetCommitRecord(
+                        request.groupId(),
+                        topic.name(),
+                        partition.partitionIndex(),
+                        offsetAndMetadata,
+                        metadataImage.features().metadataVersion()
+                    ));
+                }
+            });
+        });
+
+        return new CoordinatorResult<>(records, response);
```

Comment:
don't we need an append future for https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L395-L434?

there are some differences compared to the existing behavior such as GroupMetadata#onOffsetCommitAppend(). 

here, we blindly update the `offsetsByGroup` timeline map which simplifies the code a lot but
i'm worried this would result in unexpected behavior.

Reply:
Could you elaborate on why do you think so?

With the timeline data structure, we don't need to store the pending offsets somewhere to latter commit them when the records are committed. The timeline data structure gives us this functionality for free, isn't it? Moreover, if writing the records fails, the timeline is reverted and the request is failed. I may have missed or misunderstood something though.

Reply:
for instance, the existing behavior only updates if the group is not dead but we don't do that here. 
and in `onOffsetCommitAppend()`:
```
      if (!offsets.contains(topicPartition) || offsets(topicPartition).olderThan(offsetWithCommitRecordMetadata))
        offsets.put(topicPartition, offsetWithCommitRecordMetadata)
```

we only append if the commit time is newer whereas we don't do these checks here.

i guess checking the response status doesn't apply here since we only handle one __consumer_offsets partition per coordinator.

for pending offset commits, you're saying their use is to check whether an append is in-flight or not which is already taken care of by the timeline map right?

Reply:
Yeah, it definitely works differently.

> the existing behavior only updates if the group is not dead but we don't do that here.

In the old implementation, we have to do this because when the produce callback fires, the group may be dead so we can't store the offsets any more. In our case, we don't have to do this anymore because the timeline data structure is here, right?

> offsets(topicPartition).olderThan(offsetWithCommitRecordMetadata)

This does not check the timestamp but the append offset. This is to ensure that the offset with the largest append offset is materialized. At the moment, we don't need this because the timeline data structure guarantees the ordering. However, we may have to do something similar when we implement the transactional offsets, depending on how we do it.

> for pending offset commits, you're saying their use is to check whether an append is in-flight or not which is already taken care of by the timeline map right?

That's right. Until we commit the offset, the records are "pending" in the timeline data structure.

Reply:
thanks for the clarification.

> In the old implementation, we have to do this because when the produce callback fires, the group may be dead so we can't store the offsets any more. In our case, we don't have to do this anymore because the timeline data structure is here, right?

Isn't it "**won't** store the offsets"?

let's say that before the commit completes the group transitions to Dead. When the commit completes we replay and add the offset to `offsetsByGroup`. I'm not sure if that will cause issues but that is different from the existing behavior, no?

Reply:
> Isn't it "won't store the offsets"?

I am not sure the difference.

> let's say that before the commit completes the group transitions to Dead. When the commit completes we replay and add the offset to offsetsByGroup. I'm not sure if that will cause issues but that is different from the existing behavior, no?

This is not how it works, Jeff. The steps are the following: 1) generate records + response; 2) replay records to update the state machine (and the timeline); 3) write the records; 4) put the response to the purgatory. When the HWM advances, the request is completed in the purgatory. If 2) or 3) fail, all the changes are reverted. It is like the operation never happened.

Reply:
Ohhh we replay before we write the records.. got it. Thanks for the clarification.

Nevertheless, i still feel like we can have a case where a Dead group can have lingering committed offsets in `offsetsByGroup`

Reply:
How? Could you elaborate?

A groups transitions to Dead when all offsets are expired. This implies that offsets are deleted and the group is deleted afterwards. In your timeline data structure (based on the same snapshot registry), you will have something like this:

1) Update offset X for group Y.
2) Delete offset X (manually or through expiration)
3) Delete group Y.

The ordering is guaranteed here. 1) can't not suddenly get after 2) or 3).

Reply:
I think maybe the confusion is coming from how the old code has conditions for if the group is dead. However, the new code does not.
>  the group may be dead so we can't store the offsets any more. In our case, we don't have to do this anymore because the timeline data structure is here, right?

I think maybe Jeff is not sure how "the timeline data structure is here" prevents this. My understanding is that we can store uncommitted data there. Then we either commit or don't and the response depends on that. Is that explanation correct?

Reply:
That's correct. The key difference is that we don't have to manage uncommitted data ourselves. We can leverage the timeline data structures for this purpose.

Reply:
> A groups transitions to Dead when all offsets are expired. This implies that offsets are deleted and the group is deleted afterwards. 

I think i was missing this. This means that a group transitions to Dead only after the deletion for both group and offsets is committed right? 

To confirm, a delayed request will always be processed in the order it came in:

1. Update offset X for group Y (purgatory op1)
2. Delete offset X and group Y (purgatory op2)
3. group Y transitions to Dead. This means that purgatory op2 was already completed. Which means purgatory op1 was already committed/reverted as they are processed in order. Therefore, a group cannot be in Dead state when purgatory op1 is not yet completed.

Is my understanding correct?